src/main/java/org/zalando/nakadi/service/EventStream.java

```diff
@@ -142,30 +146,47 @@ public void streamEvents(final AtomicBoolean connectionReady) {
                 .collect(Collectors.toMap(identity(), valueFunction));
     }
 
-    public static String createStreamEvent(final Cursor cursor, final List<String> events) {
-        final StringBuilder builder = new StringBuilder()
-                .append("{\"cursor\":{\"partition\":\"").append(cursor.getPartition())
-                .append("\",\"offset\":\"").append(cursor.getOffset()).append("\"}");
+    public static int writeStreamEvent(final OutputStream outputStream, final Cursor cursor, final List<byte[]> events)
+            throws IOException {
+        final CountingOutputStream countingOutputStream = new CountingOutputStream(outputStream) {
+            @Override
+            public void flush() throws IOException {
+                // block flushing
+            }
+        };
+        final Writer writer = new OutputStreamWriter(countingOutputStream);
+        writer.write("{\"cursor\":{\"partition\":\"");
+        writer.write(cursor.getPartition());
+        writer.write("\",\"offset\":\"");
+        writer.write(cursor.getOffset());
+        writer.write("\"}");
         if (!events.isEmpty()) {
-            builder.append(",\"events\":[");
-            events.forEach(event -> builder.append(event).append(","));
-            builder.deleteCharAt(builder.length() - 1).append("]");
+            writer.write(",\"events\":[");
+            writer.flush();
+            boolean first = true;
+            for (final byte[] event : events) {
+                if (!first) {
+                    countingOutputStream.write(',');
```

Comment:
the `writer` writes to `countingOutputStream` so it's the whole size. The `writer.flush()` is called before switching to use `countingOutputStream` directly so that the internal buffers of the `writer` get flushed before that.